sigstore/_internal/sct.py

```diff
@@ -132,21 +135,41 @@ def _pack_digitally_signed(
     return data
 
 
-def _is_preissuer(issuer: Certificate) -> bool:
-    try:
-        ext_key_usage = issuer.extensions.get_extension_for_class(ExtendedKeyUsage)
-    # If we do not have any EKU, we certainly do not have CT Ext
-    except ExtensionNotFound:
-        return False
+def _get_issuer_cert(chain: List[Certificate]) -> Certificate:
+    for cert in chain:
+        if "CN=sigstore-intermediate" in [
+            name.rfc4514_string() for name in cert.subject.rdns
+        ]:
+            return cert
+    return chain[0]
```

Comment:
I think this is going to be unreliable, especially in the BYO PKI case -- the fact that the sigstore intermediate happens the have the `sigstore-intermediate` CN is probably an implementation detail 🙂 

I _think_ what we want to do here is make `_get_issuer_cert` have a signature like this:

```
def _get_issuer_cert(child: Certificate, candidates: list[Certificate]) -> Certificate
```

...and then have find the right cert in `candidates` by returning the first that has a `SubjectKeyIdentifier` that matches `child`'s `AuthorityKeyIdentifier`. Sigstore's X.509 profile should guarantee that both of those extensions are present.

Reply:
Actually, thinking a little more: an even simpler implementation of `_get_issuer_cert` is to just call [verify_directly_issued_by](https://cryptography.io/en/latest/x509/reference/#cryptography.x509.Certificate.verify_directly_issued_by) with each candidate and return the first that matches.

Reply:
It makes a lot of sense for me!! I'll just do that

Reply:
Hey William, actually I was discussing with @haydentherapper and the root cause that the `_get_issuer_cert` was not working properly was that the `chain` is not exactly a chain but a `pool` of certificates for the verifying flow. So I performed a modification on the `verify_sct` call to pass the chain and not the pool. We still have a "bug" because these pool is called as a chain the the Verifier class, what can be confused.

Reply:
Aha! Thank you for catching that -- that's always been a mild mystery to me (as to why we needed to search for the issuer at all, since it should always be the adjacent intermediate CA).

IMO we can rename the misleading parameter to `pool` or `trusted_certs` as part of this change -- both better convey that there's no "one true chain" here, and that the certs are part of a trust bundle rather than a part of a specific linear path.

Reply:
Out of scope for this PR, but another question is if the interface for [`get_fulcio_certs`](https://github.com/sigstore/sigstore-python/blob/8ea2a7c2c3664f932122a53daa67a09411beac49/sigstore/_internal/trustroot.py#L136) should be changed to return a list of chains rather than a concatenated list.

This is related to how `X509Store` works. It's not immediately clear to me how it does. It seems like you push all certificates that are trusted, intermediates or roots, into the store (which I assume is why the interface returns a single list rather than a list of lists). When we were testing, we saw the verified chain contain both the intermediate and root Fulcio certs, so I assume the store tries to verify down to a self-signed certificate? Again, unrelated to the PR, but I wanted to check that the intermediate, even if marked as trusted and not just a chain builder, is getting verified against a root.

Reply:
https://github.com/openssl/openssl/blob/master/crypto/x509/x509_vfy.c#L3204 - this is the "fun" function to read that answers this question. I _think_ openssl defines a trusted root certificate as a self-signed cert, so even if we add intermediates to the trust store, the chain must terminate with a self-signed cert. I think.

Reply:
>  I think openssl defines a trusted root certificate as a self-signed cert, so even if we add intermediates to the trust store, the chain must terminate with a self-signed cert. I think...

It does, but that's actually technically incorrect of them 😅: RFC 3280/5280 only require chain construction to the first trust anchor, regardless of whether it's a self-signed root or not. OpenSSL's default behavior is to continue searching until it hits a self-signed root in the trusted set, which caused a decent amount of pain in ~2020 (IIRC) when one of the Root CAs that cross-signed for `Let's Encrypt Authority R3` expired. They added `X509_V_FLAG_PARTIAL_CHAIN` as a flag to enable the correct behavior, but it isn't the default.

TL;DR: We might need to pass `X509_V_FLAG_PARTIAL_CHAIN` if we really do need the "long chain." But I'm not sure that we do -- isn't it sufficient to just verify against the intermediate and return the shorter chain?

Reply:
Given the root is present in the trusted root file, I would prefer we still validate it. The intermediate being shipped in the trust root is more of an optimization, letting us avoid distributing a CRL if we need to revoke the intermediate cert. The PKI root of trust is supposed to be the Fulcio root cert.

Long-term, we should revisit whether we want a Fulcio intermediate. Might make sense to just issue off the root. Originally, the thought was we'd have one Fulcio root that could be used to issue intermediates for other use cases, eg code signing and timestamping. But given we haven't needed that yet, maybe we should just issue off the root.

Reply:
> Given the root is present in the trusted root file, I would prefer we still validate it. The intermediate being shipped in the trust root is more of an optimization, letting us avoid distributing a CRL if we need to revoke the intermediate cert. The PKI root of trust is supposed to be the Fulcio root cert.

Makes sense! In that case, we can simulate the "long chain" building by passing only the Fulcio root into the trust store and including the intermediate via the `intermediates` parameter to `X509StoreContext`. That will ensure we never terminate before the self-signed TA, regardless of OpenSSL flags like `X509_V_FLAG_PARTIAL_CHAIN`.